File: packages/validation/src/purchase-order.ts
// =============================================================================
// Purchase Order schemas
// =============================================================================
import { z } from "zod";
import { uuidSchema, listQuerySchema } from "./common";

// -----------------------------------------------------------------------------
// Enums
// -----------------------------------------------------------------------------
export const poStatusSchema = z.enum(["DRAFT", "APPROVED", "POSTED", "COMPLETED", "CANCELLED"]);
export type PoStatusInput = z.infer<typeof poStatusSchema>;

export const poLineStatusSchema = z.enum(["OPEN", "PARTIAL", "RECEIVED", "CANCELLED"]);

// -----------------------------------------------------------------------------
// Lines
// -----------------------------------------------------------------------------
const numericString = (schema: z.ZodNumber) =>
  z.union([z.number(), z.string()]).transform((v, ctx) => {
    if (typeof v === "number") {
      if (!Number.isFinite(v)) {
        ctx.addIssue({ code: "custom", message: "Số không hợp lệ" });
        return z.NEVER;
      }
      return v;
    }
    const n = Number(v.replace(",", "."));
    if (!Number.isFinite(n)) {
      ctx.addIssue({ code: "custom", message: "Số không hợp lệ" });
      return z.NEVER;
    }
    return n;
  }).pipe(schema);

export const poLineInputSchema = z.object({
  product_id: uuidSchema,
  unit_id: uuidSchema,
  quantity: numericString(z.number().positive("Số lượng phải > 0")),
  unit_price: numericString(z.number().nonnegative("Đơn giá phải >= 0")),
  discount_pct: numericString(z.number().min(0).max(100)).default(0),
  tax_pct: numericString(z.number().min(0).max(1000)).default(0),
  notes: z.string().trim().max(500).optional().nullable(),
});

export type PoLineInput = z.infer<typeof poLineInputSchema>;

// -----------------------------------------------------------------------------
// Create / Update header
// -----------------------------------------------------------------------------
export const createPurchaseOrderSchema = z.object({
  branch_id: uuidSchema,
  party_id: uuidSchema,
  order_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, "Ngày phải có dạng YYYY-MM-DD"),
  expected_date: z
    .string()
    .regex(/^\d{4}-\d{2}-\d{2}$/, "Ngày phải có dạng YYYY-MM-DD")
    .optional()
    .nullable(),
  currency: z.string().length(3).default("VND"),
  exchange_rate: numericString(z.number().positive()).default(1),
  discount_amount: numericString(z.number().nonnegative()).default(0),
  shipping_amount: numericString(z.number().nonnegative()).default(0),
  payment_terms: z.coerce.number().int().nonnegative().default(0),
  shipping_address: z.string().trim().max(500).optional().nullable(),
  notes: z.string().trim().max(2000).optional().nullable(),
  internal_notes: z.string().trim().max(2000).optional().nullable(),
  // ⭐ BẮT BUỘC: Mỗi PO phải gắn với 1 hợp đồng thầu
  bid_contract_id: uuidSchema,
  bid_lot_id: uuidSchema.optional().nullable(),
  lines: z.array(poLineInputSchema).min(1, "PO phải có ít nhất 1 dòng"),
});

export const updatePurchaseOrderSchema = z.object({
  party_id: uuidSchema,
  order_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/),
  expected_date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional().nullable(),
  discount_amount: numericString(z.number().nonnegative()).optional(),
  shipping_amount: numericString(z.number().nonnegative()).optional(),
  shipping_address: z.string().trim().max(500).optional().nullable(),
  notes: z.string().trim().max(2000).optional().nullable(),
  internal_notes: z.string().trim().max(2000).optional().nullable(),
  lines: z.array(poLineInputSchema).min(1).optional(),
});

export type CreatePurchaseOrderInput = z.infer<typeof createPurchaseOrderSchema>;
export type UpdatePurchaseOrderInput = z.infer<typeof updatePurchaseOrderSchema>;

// -----------------------------------------------------------------------------
// Workflow actions
// -----------------------------------------------------------------------------
export const approvePoSchema = z.object({
  notes: z.string().trim().max(500).optional(),
});

export const cancelPoSchema = z.object({
  reason: z.string().trim().min(1, "Phải nhập lý do hủy").max(500),
});

// -----------------------------------------------------------------------------
// List query
// -----------------------------------------------------------------------------
export const poListQuerySchema = listQuerySchema.extend({
  search: z.string().trim().optional(),
  party_id: uuidSchema.optional(),
  branch_id: uuidSchema.optional(),
  status: poStatusSchema.optional(),
  date_from: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional(),
  date_to: z.string().regex(/^\d{4}-\d{2}-\d{2}$/).optional(),
});

File: apps/api/src/InventoryPro.Domain/Purchasing/PurchaseOrder.cs
namespace InventoryPro.Domain.Purchasing;

using InventoryPro.Domain.Bidding;
using InventoryPro.Domain.Catalog;
using InventoryPro.Domain.Common;
using InventoryPro.Domain.Parties;

/// <summary>
/// Trạng thái Purchase Order. Workflow:
/// DRAFT → APPROVED → POSTED → COMPLETED. Có thể CANCELLED từ bất kỳ trạng thái nào.
/// - DRAFT: tạo + sửa dòng thoải mái
/// - APPROVED: đã duyệt, không sửa dòng (chỉ có thể POST hoặc CANCEL)
/// - POSTED: đã đặt hàng với NCC, chờ GRN. Có thể nhận từng phần qua nhiều GRN
/// - COMPLETED: tất cả dòng đã nhận đủ (received_qty = quantity)
/// </summary>
public enum PurchaseOrderStatus
{
    Draft = 0,
    Approved = 1,
    Posted = 2,
    Completed = 3,
    Cancelled = 4,
}

public enum PurchaseOrderLineStatus
{
    Open = 0,
    Partial = 1,
    Received = 2,
    Cancelled = 3,
}

/// <summary>
/// Purchase Order header. Branch-scoped (mỗi PO thuộc 1 chi nhánh cụ thể).
/// party_id là supplier (Party với party_type SUPPLIER hoặc BOTH).
/// </summary>
public class PurchaseOrder : BranchScopedEntity
{
    public string PoNumber { get; set; } = string.Empty;
    public Guid PartyId { get; set; }
    public DateTime OrderDate { get; set; } = DateTime.UtcNow.Date;
    public DateTime? ExpectedDate { get; set; }
    public string Currency { get; set; } = "VND";
    public decimal ExchangeRate { get; set; } = 1;
    public decimal Subtotal { get; set; } = 0;
    public decimal DiscountAmount { get; set; } = 0;
    public decimal TaxAmount { get; set; } = 0;
    public decimal ShippingAmount { get; set; } = 0;
    public decimal Total { get; set; } = 0;
    public decimal PaidAmount { get; set; } = 0;
    public PurchaseOrderStatus Status { get; set; } = PurchaseOrderStatus.Draft;
    public int PaymentTerms { get; set; } = 0;
    public string? ShippingAddress { get; set; }
    public string? Notes { get; set; }
    public string? InternalNotes { get; set; }
    public Guid? ApprovedBy { get; set; }
    public DateTime? ApprovedAt { get; set; }
    public Guid? PostedBy { get; set; }
    public DateTime? PostedAt { get; set; }
    public DateTime? CompletedAt { get; set; }
    public DateTime? CancelledAt { get; set; }
    public string? CancelReason { get; set; }

    // FK tới hợp đồng thầu (BẮT BUỘC - validate ở handler)
    public Guid? BidContractId { get; set; }
    public Guid? BidLotId { get; set; }

    // Navigation
    public Parties.Party? Party { get; set; }
    public BidContract? BidContract { get; set; }
    public BidLot? BidLot { get; set; }
    public ICollection<PurchaseOrderLine> Lines { get; set; } = new List<PurchaseOrderLine>();
}

/// <summary>
/// Dòng chi tiết PO. Lưu product_name/unit_code tại thời điểm tạo (snapshot)
/// để tránh phụ thuộc vào việc product đổi tên/đổi đơn vị sau này.
/// received_qty được cập nhật bởi GRN handler (sẽ thêm ở module 0007).
/// </summary>
public class PurchaseOrderLine : TenantScopedEntity
{
    public Guid PurchaseOrderId { get; set; }
    public int LineNo { get; set; }
    public Guid ProductId { get; set; }
    public Guid UnitId { get; set; }
    public string ProductName { get; set; } = string.Empty;
    public string UnitCode { get; set; } = string.Empty;
    public decimal Quantity { get; set; } = 0;
    public decimal ReceivedQty { get; set; } = 0;
    public decimal UnitPrice { get; set; } = 0;
    public decimal DiscountPct { get; set; } = 0;
    public decimal TaxPct { get; set; } = 0;
    public decimal LineTotal { get; set; } = 0;
    public PurchaseOrderLineStatus Status { get; set; } = PurchaseOrderLineStatus.Open;
    public string? Notes { get; set; }

    // Navigation
    public PurchaseOrder? PurchaseOrder { get; set; }
    public Product? Product { get; set; }
    public UnitOfMeasure? Unit { get; set; }
}
